### PR TITLE
Fuse SimpleCommand and the patch modules into Command and Errors

### PR DESCRIPTION
**Pull request by Aquaj**
[[Notion]](https://www.notion.so/swile/Fusionner-nos-3-classes-dans-Command-5fcd9fb6de3443678c77eb6b1560cfef)

### CODE REVIEW

File: lib/command/errors.rb [hunk 1]
```diff
@@ -20,6 +27,21 @@ def add(attribute, code, message_or_key = code, **options)
       self[attribute].uniq!
     end
 
+    def merge_from(object)
+      raise ArgumentError unless object.respond_to?(:errors)
+      errors = if object.errors.respond_to?(:messages)
```
Comment (guicharn) [inline lib/command/errors.rb:32]: Nice ! 😃

File: lib/command.rb [hunk 2]
```diff
@@ -11,32 +11,77 @@ def initialize(message = nil, code = nil)
...
+
+  module ClassMethods
+    def call(*args)
+      new(*args).call
+    end
+
+    def extended(base)
+      base.i18n_scope = "errors.messages"
+    end
+    attr_accessor :i18n_scope
```
[the remark this answers is not in the document]
Comment (Aquaj) [reply lib/command.rb:30]: 👍  Adding this.

File: lib/command.rb [hunk 3]
```diff
@@ -11,32 +11,77 @@ def initialize(message = nil, code = nil)
...
+  end
+
+  def failure?
+    called? && errors.any?
+  end
+
+  def errors
+    @errors ||= Command::Errors.new(source: self.class)
+  end
+  alias_method :full_errors, :errors
```
[the remark this answers is not in the document]
Comment (Aquaj) [reply lib/command.rb:53]: Moved these into a module + added a deprecation warning.

File: lib/command.rb [hunk 4]
```diff
@@ -11,32 +11,77 @@ def initialize(message = nil, code = nil)
...
+
+  # Convenience/retrocompatibility aliases
+  def self.errors_alias(method, errors_method)
+    define_method method do |*args|
+      errors.send errors_method, *args
+    end
+  end
+  errors_alias :clear_errors, :clear
+  errors_alias :add_error, :add
+  errors_alias :merge_errors_from, :merge_from
```
[the remark this answers is not in the document]
Comment (Aquaj) [reply lib/command.rb]: Moved these into a module + added a deprecation warning.

File: lib/command/errors.rb (hunk 1 above)
Comment (Aquaj) [reply lib/command/errors.rb:36]: Thanks !

File: README.md [hunk 5]
```diff
@@ -344,7 +392,7 @@ end
 
 ## Matchers
 
-To simplify your life, the gem come with matchers.  
+To simplify your life, the gem come with matchers.
```
Comment (guicharn) [inline README.md:395]: This modification change the layout : it removes the carriage return

File: lib/command.rb [hunk 6]
```diff
@@ -11,32 +11,78 @@ def initialize(message = nil, code = nil)
...
+
+  def errors
+    @errors ||= Command::Errors.new(source: self.class)
+  end
+
+  module LegacyErrorHandling
+    # Convenience/retrocompatibility aliases
+    def self.errors_legacy_alias(method, errors_method)
+      define_method method do |*args|
+        puts "/!\\ #{method} is deprecated, please use errors.#{errors_method} instead."
```
Comment (guicharn) [inline lib/command.rb:58]:
```suggestion
        warn "/!\\ #{method} is deprecated, please use errors.#{errors_method} instead."
```
https://ruby-doc.org/core-2.6/Warning.html 😉

File: lib/command/errors.rb [hunk 7]
```diff
@@ -4,11 +4,23 @@ module Command
...
+
+    def initialize(source: nil)
+      @source = source
+      super()
+    end
+
+    def exists?(attribute, code)
+      self.fetch(attribute, []).any? { |e| e[:code] == code }
+    end
+    alias_method :has_error?, :exists?
```
Comment (guicharn) [inline lib/command/errors.rb:17]:
```suggestion
```

**Review by guicharn — APPROVED** [review]

File: lib/command/errors.rb (hunk 7 above)
Comment (Aquaj) [reply lib/command/errors.rb:17]: Having this allows us to do `expect(command.errors).to have_error(:attribute, :code)` natively in RSpec specs.

File: README.md (hunk 5 above)
Comment (Aquaj) [reply README.md:395]: This was auto-removed by vim. But according to GitHub's diff it doesn't remove the CR. (README in diff vs README on `master`)

Capture d’écran 2021-03-03 à 09 27 29
Capture d’écran 2021-03-03 à 09 27 36

**Pull request merged**